File: Question3/Question3.js
'use strict';

const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
});

function solution(N, M, P) {
    let is_sorted = true;
    let index = 0;
    let sum = P[index];

    for (let i = 0; i < N - 1; i++) {
        if (P[i] > P[i + 1]) {
            is_sorted = false;
            break;
        }
    }

    if (is_sorted) return 1;

    while (index < N && sum <= M) {
        for (let i = 0; i < index + 1; i++) {
            if (P[i] > P[i + 1]) {
                let temp = P[i];
                P[i] = P[i + 1];
                P[i + 1] = temp;
            }
        }

        index++;
        sum += P[index];
    }

    for (let i = 0; i < N - 1; i++) if (P[i] > P[i + 1]) return 0;

    return 1;
}

readline.question('', firstLine => {
    const [N, M] = firstLine.split(' ').map(Number);

    readline.question('', secondLine => {
        const P = secondLine.split(' ').map(Number);

        const answer = solution(N, M, P);
        console.log(answer);

        readline.close();
    });
});

File: Question3/Question3.py
def solution(N, M, P):
    is_sorted = True
    index = 0
    sum = P[index]

    for i in range(N - 1):
        if P[i] > P[i + 1]:
            is_sorted = False
            break

    if is_sorted: return 1

    while index < N and sum <= M:
        for i in range(index + 1):
            if P[i] > P[i + 1]:
                P[i], P[i + 1] = P[i + 1], P[i]

        index += 1
        sum += P[index]

    for i in range(N - 1):
        if P[i] > P[i + 1]:
            return 0

    return 1

def main():
    N, M = map(int, input().split())
    P = list(map(int, input().split()))
    answer = solution(N, M, P)
    print(answer)

if __name__ == "__main__":
    main()
